src/percent_encoding.rs

```diff
@@ -110,6 +110,14 @@ define_encode_set! {
     }
 }
 
+define_encode_set! {
+    /// This encode set is used to decide which characters are allowed in the hostname.
+    pub HOSTNAME_ENCODE_SET = [SIMPLE_ENCODE_SET] | {
```

Comment:
Encode sets are normally used for percent encoding, not decoding. So I’d rather not expose this in the public API. In `after_percent_sign` you could "inline" `!HOSTNAME_ENCODE_SET.contains(c)` to `!(matches(c, ' ' | '!' | '"' | /* ...*/ '~') || c < '\u{20}' || c > '\u{7E}')`, or invert the logic and use `matches!(c, 'a'...'z' | 'A'...'Z' | '0'...'9' | '.' | '-' | '_' | '*')`. (I think that’s the complementary set, but I’m not sure `*` should be there.)